### PR TITLE
Add featured product section

### LINKED ISSUE #36: New Section: Featured Product
**User Story**

As a merchant, I can feature a specific product through a section on any template that supports sections, so to increase visibility /discoverability of a product on my online store. 

**UI**
image

[Figma](https://www.figma.com/file/JcYdswrk2JhqBTGBVWPM7J/Dawn-2.0?node-id=2463%3A0)

### PR DESCRIPTION
**Pull request by LucasLacerdaUX**
> Note: This PR is based off #219 in order and contains the latest design updates. Make sure to rebase before merging :)

**Why are these changes introduced?**
Adds the featured product section.

Closes #36.

**What approach did you take?**
_Featured Product_
- We're reusing the same CSS and JS as product page.
- A new, separate `section-featured-product` was added with Featured Product-only code.

_Changes to other sections/scripts_
- Updated shared CustomElements so they're not defined twice if included multiple times on the page. I moved the class to the `define()` function and am now checking if the element is already defined.
- Added support for `data-update-url="false"` to toggle off updating the `window.history` when interacting with variant pickers
 
**TODOs:**
- Validate with UX Designers:
  - Final settings
  - Blocks
  - Spacing for both normal and secondary background 
  - When opening the modal, should all media be there, including ones that are NOT associated with a variant?

**Demo links**

- [Store](https://os2-demo.myshopify.com/?preview_theme_id=124633907222)
- [Editor](https://os2-demo.myshopify.com/admin/themes/124633907222/editor)

**Checklist**
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

**Comment by LucasLacerdaUX** [comment]
Reminder: 
image

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [inline sections/featured-product.liquid:13]: **Possible bug:** Using `section.settings.product` directly will result in an error with forms.



If you try to output `{{ section.settings.product.title }}`, you'll see that the ProductDrop is not empty.

Comment (tyleralsbury) [reply sections/featured-product.liquid:13]: That's weird, but I guess for now we can use the assign and bring the bug to the platform team and see if they can fix it for a refactor later.

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:8]: Is this needed ? 🤔  I don't see anything related to the slider.

File: assets/section-featured-product.css
Comment (ludoboludo) [inline assets/section-featured-product.css:1]: I think there should be more margin at the bottom when it's the last section, otherwise it gets pretty close to the footer: [screenshot](https://screenshot.click/21-08-48i5a-o3til.png)

File: assets/section-main-product.css
Comment (ludoboludo) [reply assets/section-main-product.css:293]: Oh and also wondering if we should also center the installement text and pickup availability or not: [screenshot](https://screenshot.click/21-08-itm0n-oj98s.png)

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:306]: Wondering if the styling on hover should match other links with the underline getting thicker: [video](https://screenshot.click/21-08-iw3x9-rem8t.mp4)

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:166]: out of curiosity, are hover styles for radio button going to be tackled here ? I assume it's going to be its own PR. ([figma reference](https://screenshot.click/21-08-k34r2-4gn0p.png))

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [reply sections/featured-product.liquid:8]: Good catch. It isn't :D

File: assets/section-featured-product.css
Comment (melissaperreault) [reply assets/section-featured-product.css:1]: +1

File: sections/featured-product.liquid
Comment (melissaperreault) [reply sections/featured-product.liquid:306]: +1 It should

File: sections/featured-product.liquid
Comment (melissaperreault) [reply sections/featured-product.liquid:166]: I'll need to look at this closely, the only piece I believe we were missing is the Sold out UI that would come in later.

File: assets/section-featured-product.css
Comment (LucasLacerdaUX) [reply assets/section-featured-product.css:5]: Fixed :D

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [reply sections/featured-product.liquid]: I've fixed it on the CSS side, but decided to remove the pickup availability block for now :)

**Comment by LucasLacerdaUX** [comment]
@tauthomas01 @ludoboludo I've updated the PR to address your comments. Hopefully I didn't miss any.

I've also decided to descope pickup-availability for now. Following what we decided for other features of the product page, I felt like this should be addressed separately from this section's implementation. I have absolutely no context on how this feature works, and trying to understand & fix it would only cause delays to this branch.

We can revisit it once the main feature is shipped, either by pairing with Thomas who originally implemented it or reassigning this task.

File: assets/section-main-product.css
Comment (LucasLacerdaUX) [reply assets/section-main-product.css:301]: I've updated to center these elements, following Meli's comments above that everything should be centered. Let me know if that works :D

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [reply sections/featured-product.liquid:164]: I honestly had no idea these styles existed. They're probably new 👀 

We can do it on the sold out UI PR one like meli said.

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:45]: I'm curious to know why we need both `li` logics here 🤔  I feel like I'm missing something obvious.

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:63]: I don't see the sticky setting in this section so I think we can remove this.
```suggestion
        <div id="ProductInfo-{{ section.id }}" class="product__info-container">
```

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:126]: I don't see a popup block option so I think this can be removed 🤔

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:220]: The form doesn't seem to be centered when it's no JS and no image but maybe that's too much of an edge case: 
[screenshor](https://screenshot.click/21-08-h3oiw-9bjy4.png)

File: sections/featured-product.liquid
Comment (ludoboludo) [inline sections/featured-product.liquid:60]: The button doesn't seem to take the full width on mobile: 

It's aligned to the right but not the left.

File: assets/section-featured-product.css
Comment (ludoboludo) [inline assets/section-featured-product.css:6]: Here because the media list has a negative margin on mobile, the padding-right ends up looking like too much space.

**Review by ludoboludo — COMMENTED** [review]
I find that both the featured product and the product page are lacking a bit of breathing room between blocks: [screenshot comparison with figma](https://screenshot.click/21-08-bo5wf-8bgbm.png), [screenshot on mobile](https://screenshot.click/21-08-idcpq-z4dxy.png)

The `more payment options` should use the link color and not the foreground one. [screenshot](https://screenshot.click/21-08-idcpq-z4dxy.png)

File: assets/section-featured-product.css
Comment (LucasLacerdaUX) [reply assets/section-featured-product.css:25]: That's a good suggestion IMO. I know Dawn isn't meant to be a one-page theme, but I don't see why we should stop merchants from removing or reordering this link if they want.

Let's wait for Meli on this, but I agree with both perspectives you shared.

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [reply sections/featured-product.liquid:26]: Good question. I'll check if there's any styles that are "forcing" this to be a list, but you're right, we'll only display one element at once so I don't see a reason to keep it as a `<ul>`.

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [reply sections/featured-product.liquid:45]: We're using a CSS strategy based on `:first-child` here to decide which media to display, as the variant-select scripts will always move the current selected media to the top of the list. There's no toggling of `hidden` and `visible` classes, as that isn't needed on the product page.


So, for the featured product: the first `<li>` ensures that **the first media** will always be the one attached to the currently selected variant (`selected_or_first_available_variant`). 

Consider a product with:
- Two variants. 
- Both of them have a featured media
- The first variant is sold out, so the second is the first available variant.
 
Because of that, the second variant will be selected by default. However, if we use `for media in product.media`, the first media it'll render won't be the one attached to the current variant, if that makes sense. 👀 

If you think adding a `hidden`/`visible` class logic is better, let me know. I preferred this way to avoid any modifications to the shared variant JS script.

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [reply sections/featured-product.liquid:60]: Wait. Isn't the image the one that's not correctly aligned in that case? 

Either way, I'll check it again and work on a fix 👀

Comment (ludoboludo) [reply sections/featured-product.liquid:60]: ahah good point. I noticed this first then the image last 😜

File: locales/ja.schema.json
Comment (movermeyer) [reply locales/ja.schema.json]: I think I've fixed it now.

File: sections/featured-product.liquid
Comment (ludoboludo) [reply sections/featured-product.liquid:82]: This will be fixed for the product page so I think here as well in this PR by Ken: https://github.com/Shopify/dawn/pull/407

Comment (kmeleta) [reply sections/featured-product.liquid:82]: Yeah I'm currently messing with that. Happy to handle it in my smaller PR. There's probably enough going on in this one.

File: sections/featured-product.liquid
Comment (LucasLacerdaUX) [reply sections/featured-product.liquid:157]: Good catch. Just fixed that and updated to get the latest changes from #349 that Ludo just merged. :)

**Comment by LucasLacerdaUX** [comment]
@ludoboludo @tauthomas01 all your comments have been addressed. Also rebased to get the latest changes from master.

Requesting new reviews :D

**Review by melissaperreault — COMMENTED** [review]
Thank you Lucas!!!

I really appreciate all the level of care and details from you and reviewers so far, many thanks for all the catch!

Here are some of my early observations:

**General**
- [x] `View full details` [link](https://screenshot.click/21-08-28hvz-zmp7h.png) could benefit of harmonizing the same animation from elsewhere on the theme; growing arrow without the underline [like on the blog post template](https://screenshot.click/21-08-2u5sx-0w0o6.gif), but in the other direction. (Other areas where the arrow grows horizontally are: announcement bar, predictive search, etc.)

**Default state**
- [ ] I wonder if the `Sold out` CTA could be less wide in the context of no image product? I tested 30rem and it looked more balanced. I wonder if this could be the width for when everything is centered?
- [ ] [Increase margins top and bottom](https://screenshot.click/21-08-w49o2-5o3n4.png) to space more the block button, to match the [product template spacing](https://screenshot.click/21-08-uas1n-3k8tg.png) or when a product has images.

**Answering your questions**
> Final settings + Blocks

- [ ] Featured product will have the same default blocks installed with the same datasource connected as the product template, except `Description` block. [Figma reference](https://www.figma.com/file/JcYdswrk2JhqBTGBVWPM7J/Dawn-2.0?node-id=12334%3A68218) - [Screenshot reference](https://screenshot.click/21-08-qoyjd-1j5dn.png)
- [ ] Blocks that can be added additionally to this list are: `Product rating` and `Text`
- [ ] I see custom liquid available as a block, is this new?
- [x] We would remove the description block, merchants would have other ways to add text. It's meant to be a condensed version of the product template so I would not keep it in. cc. @nicklepine 

> When opening the modal, should all media be there, including ones that are NOT associated with a variant?

- [ ] No, we would add the same setting and behaviour as for the product template from this PR https://github.com/Shopify/dawn/pull/158


**What I have left to review**
- Spacing for both normal and secondary background
- Different products and screen size
- Some accessibility checks
- Quantity UI component change still being discussed https://github.com/Shopify/dawn/issues/400

**Comment by LucasLacerdaUX** [comment]
Just pushed an update to address @melissaperreault's comments.

> I see custom liquid available as a block, is this new?

More details here: #42 

> No, we would add the same setting and behaviour as for the product template from this PR Hide variant images setting #158 

Oh, thanks for clarifying. I'll descope it from this branch, though, as the #158 isn't even merged yet for the product page context. Ideally, we should aim to get this merged quickly to get rid of rebase issues (PR is too big and been waiting for a long time to be merged), so I think this is the best decision for now.

>  Increase margins top and bottom to space more the block button, to match the product template spacing or when a product has images.

I couldn't reproduce this issue. Maybe I fixed it with any of the other changes I did. Let me know if you still see it 👀

File: sections/main-product.liquid
Comment (tauthomas01) [inline sections/main-product.liquid:36]: Could we add a documentation inside of `product-thumbnail.liquid` of what `xr_button` refers to?

File: assets/section-featured-product.css
Comment (tauthomas01) [inline assets/section-featured-product.css:45]: I think we'll need to add a "clearfix" using `:after` so we don't have end up with floating issue.

**Review by tauthomas01 — COMMENTED** [review]
After these changes, everything looks good on my end.

**Comment by Oliviammarcello** [comment]
Doing an additional review to assist @melissaperreault 
- [ ] There isn't enough spacing above the section when placed under the featured collection and collage section. [Screenshot](https://screenshot.click/21-25-5nzbl-rsysf.png)
- [ ] On mobile the space between the image and the top of the section should be changed to 25px instead of 20px. [Screenshot](https://screenshot.click/21-49-ljg6m-7osi1.png)
- [ ] The underline under `View product details` should be removed by default and only added on hover (similar to share). The arrow animation should remain. 
- [ ] When the share title is removed, the icon is misaligned with the `View product details` link. [Screenshot](https://screenshot.click/21-30-llhx8-d5gn7.png)
- [ ] When there is no product title, the content is misaligned. [Screenshot](https://screenshot.click/21-34-lat1u-01pjk.png)

**Comment by LucasLacerdaUX** [comment]
@ludoboludo @tauthomas01 Ready for new reviews. Just addressed your comments and @Oliviammarcello's design suggestions above! :)

**NEW Demo Links**
- [Store](https://os2-demo.myshopify.com/?preview_theme_id=124633907222)
- [Editor](https://os2-demo.myshopify.com/admin/themes/124633907222/editor)

File: config/settings_data.json
Comment (tauthomas01) [inline config/settings_data.json:1]: Sorry Lucas, I accidentally saved in your editor store. Would you mind reverting once the changes are approved?

File: sections/featured-product.liquid
Comment (tauthomas01) [inline sections/featured-product.liquid:18]: From [Ludo's previous comment](https://github.com/Shopify/dawn/pull/261#discussion_r691565126), this resulted with no spaces if the featured product is in first position.

If we add back the class `spaced-section`, things should be good: [video demo](https://screenshot.click/19-34-20jhp-3p1dc.mp4)

**Review by tauthomas01 — COMMENTED** [review]
One more thing I saw, then it should be good!

**Review by tauthomas01 — DISMISSED** [review]
Amazing! Thanks for doing all the changes 👏

**Review by ludoboludo — COMMENTED** [review]
- Adding a setting to show/hide `View full details ->` was mentioned [here](https://github.com/Shopify/dawn/pull/261#discussion_r691566112) but can be done in a follow up to get this big chunk out of the way. 

- I don't think we tested the share block/summary without text before. And when there is no text, like Olivia brought up, we might want to ensure that the summary is 44px by 44px and only add the `margin-right` in `.share-button__button .icon-share` when the svg isn't an `:only-child`. We would also need to ensure the svg within it is centered both vertically and horizontally. Again, could be done in follow up. 

- And then finally, in Safari, seems like we're running in some layout issue with the `View full details ->` link where the arrow goes on its own line: [screenshot](https://screenshot.click/21-08-t4o9d-4e1ls.png)

**Comment by LucasLacerdaUX** [comment]
@tauthomas01 @ludoboludo just had to rebase to fix a very small conflict on `main-product`* and that dismissed your approvals.

Can I get new ones? :DDDD

P.S.: The only change on that file was `xr_button: true` being added to `product-thumbnail`, so there's no big changes to re-review :)

**Review by tauthomas01 — APPROVED** [review]

**Review by ludoboludo — APPROVED** [review]
Looking good 👍 

In the follow up maybe we can try and see why the dynamic sources are not coming up 🤔  ([vendor](https://screenshot.click/21-08-8f4ux-z2pxb.png))

**Pull request merged**
